src/ophyd_async/core/_providers.py

```diff
@@ -17,41 +20,105 @@ class DirectoryInfo:
 
     :param root: Path of a root directory, relevant only for the file writer
     :param resource_dir: Directory into which files should be written, relative to root
-    :param prefix: Optional filename prefix to add to all files
-    :param suffix: Optional filename suffix to add to all files
+    :param filename: Base filename to use generated by FilenameProvider
+    :param create_dir_depth: Optional depth of directories to create if they do not exist
     """
 
     root: Path
     resource_dir: Path
-    prefix: Optional[str] = ""
-    suffix: Optional[str] = ""
+    filename: Optional[str] = ""
+    create_dir_depth: Optional[int] = 0
+
 
+class FilenameProvider(Protocol):
+
+    @abstractmethod
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        """Get a filename to use for output data"""
 
 class DirectoryProvider(Protocol):
+
+    _filename_provider: FilenameProvider
+    _prefix: Optional[str]
+    _suffix: Optional[str]
+
     @abstractmethod
-    def __call__(self) -> DirectoryInfo:
+    def __call__(self, device_name=None) -> DirectoryInfo:
         """Get the current directory to write files into"""
 
 
+class StaticFilenameProvider(FilenameProvider):
+    def __init__(self, filename):
+        self._static_filename = filename
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{self._static_filename}{suffix}'
+
+
+class DeviceNameFilenameProvider(FilenameProvider):
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{device_name}{suffix}'
+
+
+class UUIDFilenameProvider(FilenameProvider):
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{uuid.uuid4()}{suffix}'
+
+
 class StaticDirectoryProvider(DirectoryProvider):
     def __init__(
         self,
+        filename_provider: FilenameProvider,
         directory_path: Union[str, Path],
         filename_prefix: str = "",
         filename_suffix: str = "",
         resource_dir: Path = Path("."),
+        create_dir_depth: int = 0,
     ) -> None:
+        self._filename_provider = filename_provider
+        self._directory_path = directory_path
         if isinstance(directory_path, str):
-            directory_path = Path(directory_path)
-        self._directory_info = DirectoryInfo(
-            root=directory_path,
-            resource_dir=resource_dir,
-            prefix=filename_prefix,
-            suffix=filename_suffix,
+            self._directory_path = Path(directory_path)
+        self._resource_dir = resource_dir
+        self._prefix = filename_prefix
+        self._suffix = filename_suffix
+        self._create_dir_depth = create_dir_depth
+
+    def __call__(self, device_name=None) -> DirectoryInfo:
+        return DirectoryInfo(
+            root=self._directory_path,
+            resource_dir=self._resource_dir,
```

Comment:
@DiamondJoseph do we need control about which bit of the path is root and which bit is the resource dir?

Reply:
Currently we're writing root = /data/YYYY/visit, resource_dir = ".", but we probably want it to be resource_dir=dataCollection imminently, so all our h5s are neatly in a folder with the same name as the nexus file they are linked from.

Reply:
This is probably the job of the directory provider then to decide which is root and which is resource dir

Reply:
Had a conversation with @danielballan and @skarakuzu who say that root is going away, so I suggest we don't expose it in the `PathProvider` interface at all, then hardcode it to `/` in the old `StreamResource` until June when we can get rid of it completely

Reply:
Should I keep it as `resource_dir` or rename to `resource_path`?

Reply:
I oppose that change: our NeXus writing assumes that resource_dir is a relative path from the visit directory to do relative linking of h5 files.